id.go

```diff
@@ -68,11 +74,109 @@ func ParseDeviceID(deviceName string) (DeviceID, error) {
 		return invalidDeviceID, ErrorInvalidDeviceName
 	}
 
-	var (
-		prefix = strings.ToLower(match[1])
-		idPart = match[2]
-	)
+	return makeDeviceID(match[1], match[2])
+
+}
+
+// Locator represents a device locator, which is a device identifier an optional
+// service name and an optional ignored portion at the end.
+//
+// The general format is:
+//
+//	{scheme}:{authority}/{service}/{ignored}
+type Locator struct {
+	// Scheme is the scheme type of the locator.  A CPE will have the forms
+	// `mac`, `uuid`, `serial`.  A server or cloud service will have the form
+	// `dns`.  An event locator that is used for pub-sub listeners will have
+	// the form `event`.
+	//
+	// The Scheme MUST NOT be used to determine where to send a message, but
+	// rather to determine how to interpret the authority and service.
+	//
+	// The Scheme value will always be lower case.
+	Scheme string
+
+	// Authority is the authority portion of the locator.  For a CPE, this
+	// will be the device identifier.  For a server or cloud service, this
+	// will be the DNS name of the service.  For an event locator, this will
+	// be the event name.
+	Authority string
+
+	// Service is the service name portion of the locator.  This is optional
+	// and is used to identify which service(s) the message is targeting or
+	// originated from.  A Service value will not contain any `/` characters.
+	Service string
+
+	// Ignored is an optional portion of the locator that is ignored by the
+	// WRP spec, but is provided to consumers for their usage.  The Ignored
+	// value will contain a prefix of the `/` character.
+	Ignored string
+
+	// id is the device identifier portion of the locator if it is one.
+	id *DeviceID
+}
+
+// ParseLocator parses a raw locator string into a canonicalized locator.
+func ParseLocator(locator string) (*Locator, error) {
+	match := LocatorPattern.FindStringSubmatch(locator)
+	if match == nil {
+		return nil, ErrorInvalidLocator
+	}
+
+	var l Locator
+
+	l.Scheme = strings.ToLower(match[1])
+	l.Authority = match[2]
+	if len(match) > 3 {
+		l.Service = strings.TrimPrefix(match[3], "/")
+	}
+	if len(match) > 4 {
+		l.Ignored = match[4]
+	}
+
+	switch l.Scheme {
+	case "mac", "uuid", "serial": // device_id locators
+		id, err := makeDeviceID(l.Scheme, l.Authority)
+		if err != nil {
+			return nil, err
+		}
+		l.id = &id
+	default:
+	}
+
+	return &l, nil
+}
+
+// DeviceID returns the device identifier portion of the locator.
+func (l *Locator) DeviceID() *DeviceID {
```

Comment:
Since these methods don't change the state of the struct, I'd make take value receivers rather than pointer receivers.